Change in RenderingDevice decode api and docs

ChristopheClaustre opened the pull request:
**My goal here is** to add some method on RenderingDevice for ease of use in godot script. As it is easy to push a structure in c++ on GPU it's not as easy in godot script so I just made some method to be able to push/retrieve a PackedVector2Array, a PackedColorArray, a PackedFloat32Array... I think it is useful and I am actually using it for compute shaders ;)

**RenderingDevice::buffer_update:**
It is now two template functions.
One to give an array of element and one to give only one element for ease of use.

Those two template functions are exposed as specialized functions in gdscript for the different type of array and their type T.

**PackedByteArray::to_XXX_array:**
Became a template function to avoid code duplication and then the content of this function have been moved to a Vector<T> ctor to make it more widely available.
Add three more specialization of this method for vector2, vector3 and color as it is commonly used type in shader.

Discussion started here but nobody seems to have something to say against it : https://github.com/godotengine/godot-proposals/issues/3042

Previous PR & godot-proposals linked to this issues :
https://github.com/godotengine/godot-proposals/issues/2993
https://github.com/godotengine/godot/pull/50535

ChristopheClaustre committed:
PackedByteArray: to_XXX_array is now a template

ChristopheClaustre committed:
PackedByteArray: add of to_vector2/vector3/color_array

ChristopheClaustre committed:
rendering_device: Add of a private virtual _buffer_get_data(...) to fetch uint8_t * that is used by the now templated RenderingDevice::buffer_get_data(...)
rendering_device: New template method buffer_update<T> to upload a Vector<T> on the GPU
rendering_device: New methods buffer_update_generic, buffer_get_data_generic and uniform/storage_buffer_create_generic that takes a Variant (or a Variant::Type) as parameter instead of PackedByteArray
rendering_device: new methods _buffer_get_size to get the size of a buffer

ChristopheClaustre committed:
Docs for PackedByteArray and RenderingDevice

ChristopheClaustre commented:
The main error is:
`'framebuffer_get_format' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]`
With different functions involved.

I suppose that the error has not been found before because the file has not been modified since a while.
Should I correct it even if I didn't altered the involved functions ?

clayjohn commented:
> I suppose that the error has not been found before because the file has not been modified since a while.
Should I correct it even if I didn't altered the involved functions ?

Yes, please correct it, but in a separate PR.

ChristopheClaustre commented:
Fellow devs, this PR need review ! :)

ChristopheClaustre commented:
With some code examples, the gain will be more visible.

### With Godot's master:
```
var vectors : PackedVector2Array
# fill vectors
rd.buffer_update_vector2(buffer_vectors, 0, vectors.size() * 8, vectors.to_byte_array())

var mouse_x : PackedFloat32Array = [ get_global_mouse_position().x ]
rd.buffer_update_float32_element(buffer_mouse, 0, 4, mouse_x.to_byte_array())
# or
var mouse_y : PackedByteArray = []
mouse_y.resize(4)
mouse_y.encode_float(get_global_mouse_position().y)
rd.buffer_update_float32_element(buffer_mouse, 4, 4, mouse_y)
```
⚠️ Reminder : to_byte_array() -> mem_cpy()

### With this PR:
```
var vectors : PackedVector2Array
# fill vectors
rd.buffer_update_vector2(buffer_vectors, 0, vectors.size(), vectors)
rd.buffer_update_float32_element(buffer_mouse, 0, get_global_mouse_position().x)
rd.buffer_update_float32_element(buffer_mouse, 1, get_global_mouse_position().y)
```

ChristopheClaustre commented:
@clayjohn @Chaosus This PR need reviews guys.
I talked with @reduz and @clayjohn on the contributor chat and they are discussing the interest of `RenderingDevice::buffer_update_XXX_element` methods. They think that it may not be a good idea to encourage to send a buffer element by element (it was not my intention and not the purpose of this method but effectively it may encourage that).

clayjohn reviewed (commented):
This PR looks mostly good!

I would remove ``buffer_update_element`` and rename ``buffer_update_array`` to ``buffer_update``. Updating a single element at a time should be strongly discouraged and can lead to problems if users call buffer_update too frequently. 

I am concerned about adding to the bloat in RenderingDevice when we could likely extend GDScript organically enough to pass buffers as bytearrays. That being said, I do like how easy this makes it to update buffers from any language. 

I'm also unsure about the changes to ``buffer_get_data``. We try to avoid using lambdas as much as possible and I don't really see the benefit. Aren't you still forced to copy the memory over anyway?

ChristopheClaustre commented:
As discussed on the contributor chat, I replaced the `buffer_get_XXX_data` and `buffer_update_XXX` by `buffer_get_data_generic` and `buffer_update_generic` that takes a `Variant::Type` or a `Variant`.

I also removed the `buffer_update_XXX_element` because it encourages to push buffers element by element and it is not recommended to proceed like that.

I also added a `uniform_buffer_create_generic` and `storage_buffer_create_generic` that takes a `Variant`.
But I didn't added a generic version of `texture_buffer_create` as it works with a "format", I was not sure it will be useful to have it.

clayjohn commented on servers/rendering/rendering_device.cpp:
```diff
@@ -242,8 +244,30 @@ RID RenderingDevice::_uniform_set_create(const Array &p_uniforms, RID p_shader,
+	uint8_t *data = nullptr;
+	_buffer_get_data(p_buffer, &data, [&buffer_data](size_t p_size) {
+		Error ok = buffer_data.resize(p_size / sizeof(T));
+		assert(ok == Error::OK);
```
We prefer to use our own error and warning macros defined here https://github.com/godotengine/godot/blob/master/core/error/error_macros.h

ChristopheClaustre replied:
oh okay ^^" will fix that so

reduz replied:
We have a no-assert policy, everything has to be recoverable. just ERR_FAIL* with an error.

ChristopheClaustre replied:
Okay, good to know !

clayjohn commented on servers/rendering/rendering_device.cpp:
```diff
@@ -242,8 +335,57 @@ RID RenderingDevice::_uniform_set_create(const Array &p_uniforms, RID p_shader,
+Vector<T> RenderingDevice::buffer_get_data(RID p_buffer) {
+	Vector<T> buffer_data;
+	uint8_t *data = nullptr;
+	_buffer_get_data(p_buffer, &data, [&buffer_data](size_t p_size) {
```
@reduz will have to decide if he wants to allow a lambda here. As far as I know, the general rule is not to use them, but reduz can approve on a case-by-case basis.

ChristopheClaustre replied:
I can shot a bullet in my own foot by saying that maybe we can have a method to retrieve the buffer size to be able to initialize the `void *` before calling `buffer_get_data`. I didn't check the feasibility of this solution.

reduz replied:
No lambda, just write to a void* and you are supposed to know the size and have pre allocated it.

ChristopheClaustre commented on servers/rendering/rendering_device.cpp:
```diff
@@ -242,8 +335,57 @@ RID RenderingDevice::_uniform_set_create(const Array &p_uniforms, RID p_shader,
+	return buffer_data;
+}
+
+Variant RenderingDevice::buffer_get_data_generic(RID p_buffer, Variant::Type p_type) {
```
My preference goes for the variant solution, because the other solution is filling way too much the documentation with those functions.

reduz commented on servers/rendering/rendering_device.cpp:
```diff
@@ -242,8 +242,16 @@ RID RenderingDevice::_uniform_set_create(const Array &p_uniforms, RID p_shader,
+	size_t size = sizeof(T);
+	return buffer_update(p_buffer, p_index * size, p_size * size, (uint8_t *)p_data.ptr(), p_post_barrier);
+}
+
```
As I mentioned, updating an element should be discouraged, its not even possible with many of the buffer types due to granularity.

reduz commented on servers/rendering/rendering_device.cpp:
```diff
@@ -404,7 +412,24 @@ void RenderingDevice::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("buffer_update_float64", "buffer", "offset", "size", "data", "post_barrier"), &RenderingDevice::_buffer_update_array<double>, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("buffer_update_vector2", "buffer", "offset", "size", "data", "post_barrier"), &RenderingDevice::_buffer_update_array<Vector2>, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("buffer_update_vector3", "buffer", "offset", "size", "data", "post_barrier"), &RenderingDevice::_buffer_update_array<Vector3>, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("buffer_update_color", "buffer", "offset", "size", "data", "post_barrier"), &RenderingDevice::_buffer_update_array<Color>, DEFVAL(BARRIER_MASK_ALL));
```
I would remove the Vector3 version because alignment is not possible, so it will never be used.

reduz commented on servers/rendering/rendering_device.cpp:
```diff
@@ -404,7 +412,24 @@ void RenderingDevice::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("buffer_update_color", "buffer", "offset", "size", "data", "post_barrier"), &RenderingDevice::_buffer_update_array<Color>, DEFVAL(BARRIER_MASK_ALL));
+
+	ClassDB::bind_method(D_METHOD("buffer_update_element", "buffer", "offset", "data", "post_barrier"), &RenderingDevice::_buffer_update_element<uint8_t>, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("buffer_update_int32_element", "buffer", "offset", "data", "post_barrier"), &RenderingDevice::_buffer_update_element<int32_t>, DEFVAL(BARRIER_MASK_ALL));
```
As mentioned before, elements should not be updateable because of granularity limitations, its never a good practice.

reduz commented on drivers/vulkan/rendering_device_vulkan.h:
```diff
@@ -744,6 +744,8 @@ class RenderingDeviceVulkan : public RenderingDevice {
 
 	RID_Owner<UniformSet, true> uniform_set_owner;
 
+	virtual uint64_t _buffer_get_data(RID p_buffer, uint8_t **p_mem, const std::function<void *(size_t)> &p_memcpy = malloc) override; //this causes stall, only use to retrieve large buffers for saving
```
This function should not be virtual, if the syntax is changed to write to a pointer, it should write to a pointer, and you should know the size. Then you can add the script helpers to get as other data types.

reduz commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -5970,7 +5970,7 @@ Error RenderingDeviceVulkan::buffer_clear(RID p_buffer, uint32_t p_offset, uint3
 }
 
-Vector<uint8_t> RenderingDeviceVulkan::buffer_get_data(RID p_buffer) {
+uint64_t RenderingDeviceVulkan::_buffer_get_data(RID p_buffer, uint8_t **p_mem, const std::function<void *(size_t)> &p_malloc) {
```
Not really happy with this, I get the intention but its better you pass the pointer to where to write and dont allocate. You are supposed to know the size of this, so returning it is not necesary. Feel free to add a buffer_get_size if you want to query it.

reduz commented on core/templates/vector.h:
```diff
@@ -253,6 +250,13 @@ class Vector {
 
 	_FORCE_INLINE_ Vector() {}
 	_FORCE_INLINE_ Vector(const Vector &p_from) { _cowdata._ref(p_from._cowdata); }
+	_FORCE_INLINE_ Vector(const uint8_t *p_from, int32_t p_size) {
```
I don't think this function is a good idea, will never be used as often to justify it. Just memcpy manually to it.

reduz commented:
If I have to be entirely and completely honest, I still think that this PR adds more bloat than proper use and feel inclined to not merge it. My feeling is that its easier to just bind a void* function, worst case, to the new binder so if you use from an extension you dont have to go through PackedArray, but using these kind of things from GDScript and expecting optimization seems far fetched.

Let me think more time about it, because I am still unconvinced.

reduz commented:
To Clarify I thought initially this PR was just about adding a unique single buffer_update function that took Variant as buffer argument, and then called buffer_update depending on whether any of the types contained was valid as a buffer and error otherwise. This was a simple change, so to me its fine. 

This PR now ends up to touching too many things and making the code more complex for something that is not warranted to be an actual need from users.

reduz commented:
I suggest this API instead, which does minimum changes to the driver.

```C++
//instead of
virtual Vector<uint8_t> buffer_get_data(RID p_buffer) = 0; //this causes stall, only use to retrieve large buffers for saving
saving
//
virtual uint64_t buffer_get_size(RID p_buffer) const = 0;
virtual void buffer_write_data(RID p_buffer,uint64_t p_offset,uint64_t p_size,void *p_ptr) = 0; //not exposed to binder
Vector<uint8_t> buffer_get_data(RID p_buffer); //exposed to binder, but not virtual
Variant buffer_get_data_var(RID p_buffer,Variant::Type); //exposed to binder, but not virtual
//also add
void buffer_update_var(RID p_buffer,uint64_t p_offset, uint64 p_size, const Variant& p_data); //not virtual
```

ChristopheClaustre commented:
@reduz I think you are right, I may have been carried away while coding ^^"
It's a good argument that sending/retrieving data on GPU is indeed slow and one memcpy will not make a huge difference.
I will try to rewrite the PR to have modification only in the base class RenderingDevice.

mhilbrunner commented:
I will close this for now due to the above discussion. Thanks for contributing :)
Feel free to open a new PR once you reworked it (and add a link to this one in the description), or ping us to open this one.
